File: app/src/main/java/me/instabattle/app/managers/UserManager.java
package me.instabattle.app.managers;

import java.util.List;

import me.instabattle.app.models.Token;
import me.instabattle.app.models.User;
import me.instabattle.app.settings.State;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.http.Body;
import retrofit2.http.GET;
import retrofit2.http.POST;
import retrofit2.http.PUT;
import retrofit2.http.Path;
import retrofit2.http.Query;

public class UserManager {
    private static final String TAG = "UserManager";
    private static final UserService service = ServiceGenerator.createService(UserService.class);
    private static UserService tokenService;

    static void initTokenService() {
        tokenService = ServiceGenerator.createService(UserService.class, State.token);
    }

    public static void getAndDo(Integer userId, Callback<User> callback) {
        service.get(userId).enqueue(callback);
    }

    public static void getAndDo(String username, Callback<User> callback) {
        service.get(username).enqueue(callback);
    }

    private static void getCountAndDo(Integer count, Callback<List<User>> callback) {
        service.getCount(count).enqueue(callback);
    }

    public static void getAllAndDo(Callback<List<User>> callback) {
        getCountAndDo(null, callback);
    }

    public static void updateAndDo(Integer userId, User user, Callback<User> callback) {
        tokenService.update(userId, user).enqueue(callback);
    }

    public static void createAndDo(String username, String email, String password, Callback<User> callback) {
        service.create(new User(username, email, password)).enqueue(callback);
    }

    public static void getTokenAndDo(String email, String password, Callback<Token> callback) {
        UserService loginService = ServiceGenerator.createService(UserService.class,
                email, password);
        loginService.getToken().enqueue(callback);
    }

    interface UserService {
        @GET("users/{user_id}")
        Call<User> get(@Path("user_id") Integer userId);

        @GET("users/{username}")
        Call<User> get(@Path("username") String username);

        @GET("users/")
        Call<List<User>> getCount(@Query("count") Integer count);

        // FIXME
        @PUT("users/{user_id}")
        Call<User> update(@Path("user_id") Integer userId, @Body User user);

        @POST("users/")
        Call<User> create(@Body User user);

        @GET("token")
        Call<Token> getToken();
    }
}

File: app/src/main/kotlin/me/instabattle/app/managers/UserManager.kt
package me.instabattle.app.managers

import io.reactivex.Single
import me.instabattle.app.models.Token
import me.instabattle.app.models.User
import me.instabattle.app.settings.KState
import me.instabattle.app.settings.State
import retrofit2.Call
import retrofit2.Callback
import retrofit2.http.*

object UserManager {
    private val TAG = "UserManager"
    private val service = ServiceGenerator.createService(UserService::class.java)
    private var tokenService: UserService? = null

    fun initTokenService() {
        tokenService = ServiceGenerator.createService(UserService::class.java, KState.token)
    }

    fun getAndDo(userId: Int?, callback: Callback<User>) {
        val call = service.get(userId)
        call.enqueue(callback)
    }

    fun getAndDo(username: String, callback: Callback<User>) {
        val call = service.get(username)
        call.enqueue(callback)
    }

    fun getCountAndDo(count: Int?, callback: Callback<List<User>>) {
        val call = service.getCount(count)
        call.enqueue(callback)
    }

    fun getAllAndDo(callback: Callback<List<User>>) {
        getCountAndDo(null, callback)
    }

    fun updateAndDo(userId: Int?, user: User, callback: Callback<User>) {
        val call = tokenService!!.update(userId, user)
        call.enqueue(callback)
    }

    fun createAndDo(username: String, email: String, password: String, callback: Callback<User>) {
        val call = service.create(User(username, email, password))
        call.enqueue(callback)
    }

    fun getTokenAndDo(email: String, password: String, callback: Callback<Token>) {
        val loginService = ServiceGenerator.createService(UserService::class.java,
                email, password)
        val call = loginService.token
        call.enqueue(callback)
    }

    fun create(username: String, email: String, password: String): Single<User> =
            service.createPromise(User(username, email, password))

    fun getToken(email: String, password: String): Single<Token> =
        ServiceGenerator.createService(UserService::class.java, email, password).promiseToken

    fun get(id: Int): Single<User> = service.getPromise(id)

    fun get(username: String): Single<User> = service.getPromise(username)

    internal interface UserService {
        @get:GET("token")
        val promiseToken: Single<Token>

        @POST("users/")
        fun createPromise(@Body user: User): Single<User>

        @GET("users/{user_id}")
        fun getPromise(@Path("user_id") userId: Int?): Single<User>

        @GET("users/{username}")
        fun getPromise(@Path("username") username: String): Single<User>

        @get:GET("token")
        val token: Call<Token>

        @GET("users/{user_id}")
        fun get(@Path("user_id") userId: Int?): Call<User>

        @GET("users/{username}")
        fun get(@Path("username") username: String): Call<User>

        @GET("users/")
        fun getCount(@Query("count") count: Int?): Call<List<User>>

        // FIXME
        @PUT("users/{user_id}")
        fun update(@Path("user_id") userId: Int?, @Body user: User): Call<User>

        @POST("users/")
        fun create(@Body user: User): Call<User>
    }
}